### PR TITLE
Umcs 419/update basket

### PR DESCRIPTION
**Pull request by Ryouzanpaku**
- Update basket resource to support v2/basket.

### CODE REVIEW

File: test/integration/BasketTest.php [hunk 1]
```diff
@@ -241,6 +244,291 @@ public function chargeTransactionsShouldCreateBasketIfItDoesNotExistYet(): void
...
+        $fetchedBasket = $this->unzer->fetchBasket($basket->getId());
+        $this->assertEquals($basket->expose(), $fetchedBasket->expose());
+    }
+
+    /**
+     * Verify basket can be created and fetched.
+     *
+     * @test
+     */
+    public function maxV2BasketShouldBeCreatableAndFetchableWorkAround(): void
```
Comment (sixer1182) [inline test/integration/BasketTest.php:312]: please describe the "workaround" in the comment of the function

File: test/BasePaymentTest.php [hunk 2]
```diff
@@ -199,6 +199,28 @@ public function createBasket(): Basket
...
+    /**
+     * Creates a v2 Basket resource and returns it.
+     *
+     * @return Basket
+     */
+    public function createV2Basket(): Basket
+    {
+        $orderId = 'b' . self::generateRandomId();
+        $basket = new Basket($orderId);
+        $basket->setTotalValueGross(99.99)
```
Comment (sixer1182) [inline test/BasePaymentTest.php:211]: I could not find a test which proves that basket v2 will be used, if old and new parameters are set.
Is this tested somehow?

Comment (sixer1182) [reply test/BasePaymentTest.php:211]: now i found it

**Review by sixer1182 — COMMENTED** [review]
please check my comments

File: test/integration/BasketTest.php (hunk 1 above)
Comment (Ryouzanpaku) [reply test/integration/BasketTest.php:312]: This test doesnt actually test the workaround. So i will remove it from the functions name. Of cause I will still add the comment to the "fetschBasket" funktion.

File: src/Resources/EmbeddedResources/BasketItem.php [hunk 3]
```diff
@@ -37,19 +37,45 @@ class BasketItem extends AbstractUnzerResource
...
+    /**
+     * @var float $amountDiscount
+     *
+     * @deprecated 1.1.5.0 @see $amountDiscountPerUnitGross.
+     */
     protected $amountDiscount = 0.0;
 
-    /** @var float $amountGross */
+    /** @var float $amountDiscountPerUnitGross */
+    protected $amountDiscountPerUnitGross = 0.0;
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/EmbeddedResources/BasketItem.php:48]: Not sure which I should follow then. To make it consistent with the other existing properties, I would consider keep these not nullable, but maybe we should discuss that.

File: src/Resources/AbstractUnzerResource.php [hunk 4]
```diff
@@ -160,6 +160,16 @@ public function setSpecialParams(array $specialParams): self
...
     }
 
+    /**
+     * Returns the API version for this resource.
+     *
+     * @return string
+     */
+    public function getApiVersion(): string
+    {
+        return Unzer::API_VERSION;
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/AbstractUnzerResource.php:170]: Added an assertion to check version in abstract class unit test now.

File: src/Resources/Basket.php [hunk 5]
```diff
@@ -27,19 +27,35 @@
...
+     *
+     * @deprecated 1.1.5.0 @see $totalValueGross.
+     */
     protected $amountTotalGross = 0.0;
 
-    /** @var float $amountTotalDiscount */
+    /**
+     * @var float $amountTotalDiscount
+     *
+     * @deprecated 1.1.5.0 Property is redundant and is no longer needed.
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/Basket.php]: Okay, refered to the property of basketItem that could would be relevant to replace this.

File: src/Resources/Basket.php [hunk 6]
```diff
@@ -27,19 +27,35 @@
...
+     */
     protected $amountTotalDiscount = 0.0;
 
-    /** @var float $amountTotalVat */
+    /**
+     * @var float $amountTotalVat
+     *
+     * @deprecated 1.1.5.0 Property is redundant and is no longer needed.
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/Basket.php]: Okay, refered to the property of basketItem that could would be relevant to replace this.

File: src/Resources/EmbeddedResources/BasketItem.php [hunk 7]
```diff
@@ -187,6 +223,46 @@ public function setAmountGross(float $amountGross): BasketItem
...
+    {
+        return $this->amountDiscountPerUnitGross;
+    }
+
+    /**
+     * @param float $amountDiscountPerUnitGross
+     *
+     * @return BasketItem
+     */
+    public function setAmountDiscountPerUnitGross(float $amountDiscountPerUnitGross): BasketItem
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/EmbeddedResources/BasketItem.php:236]: As discussed, this is not needed anymore.

File: src/Resources/EmbeddedResources/BasketItem.php [hunk 8]
```diff
@@ -187,6 +223,46 @@ public function setAmountGross(float $amountGross): BasketItem
     /**
      * @return float
      */
+    public function getAmountDiscountPerUnitGross(): float
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/EmbeddedResources/BasketItem.php:226]: As discussed, this is not needed anymore.

File: src/Resources/EmbeddedResources/BasketItem.php [hunk 9]
```diff
@@ -37,19 +37,45 @@ class BasketItem extends AbstractUnzerResource
...
-    /** @var float $amountDiscount */
+    /**
+     * @var float $amountDiscount
+     *
+     * @deprecated 1.1.5.0 @see $amountDiscountPerUnitGross.
+     */
     protected $amountDiscount = 0.0;
 
-    /** @var float $amountGross */
+    /** @var float $amountDiscountPerUnitGross */
```
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/EmbeddedResources/BasketItem.php:47]: As discussed, this is not needed anymore.

File: src/Resources/EmbeddedResources/BasketItem.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ryouzanpaku) [reply src/Resources/EmbeddedResources/BasketItem.php:48]: As discussed, this is not needed anymore.

**Review by sixer1182 — APPROVED** [review]
LGTM, thanks

**Pull request merged**